in-progress/world-state/0004-world-state.md

```diff
@@ -0,0 +1,193 @@
+2. Block 2 appends a 3rd node. New nodes are added to the store along with new values for nodes that have changed. In this case, the root and the left child of the root have been updated.
+3. Block 3 updates leaf 0 causing 4 new nodes to be created up the left hand side of the tree.
+
+![image](./snapshot-tree-structure.png)
+
+Reading the sibling path for leaf 3 at block 2 is performed by traversing the tree from root to leaf, starting from the root node as located in the index. Taking the block 2 update as the starting point as that contains the appropriate root value, we can walk down the tree using the values stored at each node to determine the correct child node. In this diagram, the orange nodes are empty and are indicated as such by the absence of part of the value in the parent's key value pair. Crucially, node with value 0x6195 is found despite being written during the update of block 1, and remaining constant during block 2's update.
+
+![image](./historic-hash-path.png)
+
+This system of content addressing is used for snapshotting both append-only and indexed trees. Block 3 updated leaf 0, which could only happen in an indexed tree and it should be clear that the same method of hash path retrievel works in this case. It enables us to serve merkle membership requests for any block in time whilst only storing the changes that occur with each block.
+
+Despite this method of only storing changes with each block. Historical trees will still require a signifcant amount of data and as it stands there is no ability to prune the history meaning nodes either store all history or no history.
+
+## Updates to Block Building
+
+Though still in the design stages, we are moving towards a system of block-building where there will inevitably be 2 chains. The first being a `finalized` chain that is considered immutable. The second will have less certainty over its finality, meaning we have to consider its state updates to be susceptible to re-org or rolling back. We will refer to this as the `pending` chain. It is likely that the pending chain will become `finalized` in batches as large rollups are produced. That is to say that it it unlikely blocks will be finalized individually but as so-called `epochs` of say 32 blocks.
+
+## Images
+
+Taking the 3-stage (uncommitted, committed, snapshotted) state architecture outlined above, we can make the concept generic and introduce the term `image` representing a cahe of 'pending' state accumulated through a selection of blocks on top of the `finalized` chain state. The world state module will create an `image` for every tree at every block on the pending chain as it is read from L1. Additionally, images can be requested on demand for block-building activities and later destroyed. Images that extend beyond the previous pending epoch will not duplicate the state of that epoch. Instead they will reference the image at the tip of that epoch. All images will reference the `finalized` store.
```

Comment:
> as it is read from L1

Perhaps we also want to enable nodes to 'optimistically' create images from not-yet-proposed-to-L1 blocks that they've seen in the mempool. Even if a proposal hasn't been finalised on L1, a user might wish to act on their view of the mempool. E.g. the next sequencer might take a risk and make a head start on preparing their block, or an (advanced) user might wish to optimistically begin generating a tx from the state root of the not-yet-proposed-to-L1 block.

Reply:
Yes, a sequencer who is next in line will most definitely want to do this. If they wait until the prior proposal is confirmed on L1 they will only have a short period to build their own proposal.